### PR TITLE
feat: API gateway for the One Platform

### PR DESCRIPTION
**Pull request by riginoommen**
# Resolves ONEPLAT-205, ONEPLAT-105, ONEPLAT-204, ONEPLAT-208, ONEPLAT-209

### 

# Explain the feature
API Gateway for the One Platform which creates a unique bridge for multiple microservices. and creates a single endpoint for multiple microservices

## Does this PR introduce a breaking change
Yes

## Screenshots
N/A

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [x] Does the change have appropriate unit tests?
- [x] Did tests pass?
- [x] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [x] Was this feature demo'd and the design review approved?

### CODE REVIEW

**Comment by riginoommen** [comment]
Dependent PR - https://github.com/1-Platform/one-platform/pull/199

File: packages/api-gateway-service/.env.example [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+API=https://localhost:4000
+GRAPHQL_API=https://localhost:4000/graphql
+
+## Certificate Configuration
+## Copy the SSL Certificate and Key path to the SSL_KEY variable and SSL_CERT directory.
+# SSL_KEY=
+# SSL_CERT=
+
+## Cookie Parser
+COOKIE_PARSER=API-GATEWAY_SECRET
```
Comment (deshmukhmayur) [inline packages/api-gateway-service/.env.example:17]:
```suggestion
COOKIE_PARSER=
```

File: packages/api-gateway-service/.env.example [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+## Authenication
```
Comment (deshmukhmayur) [inline packages/api-gateway-service/.env.example:19]:
```suggestion
## Authentication
```

File: packages/api-gateway-service/package.json [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  "name": "api-gateway",
+  "description": "API Gateway for One Platform",
+  "opservice": {
+    "className": "ApiGateway",
+    "serviceName": "api-gateway"
+  },
+  "version": "0.0.1",
+  "keywords": [
+    "web-components",
+    "html"
```
Comment (deshmukhmayur) [inline packages/api-gateway-service/package.json:11]: Are you using web-components and html in this package?

File: packages/api-gateway-service/package.json [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  ],
+  "scripts": {
+    "build:dev": "webpack --config webpack.dev.js",
+    "start:dev": "webpack --watch --config webpack.dev.js",
+    "build": "webpack --config webpack.prod.js",
+    "start": "webpack --watch --config webpack.prod.js"
+  },
+  "author": {
+    "name": "Rigin Oommen",
+    "email": "riginoommen@gmail.com"
```
Comment (deshmukhmayur) [inline packages/api-gateway-service/package.json:33]: Use your red hat email here.

File: packages/api-gateway-service/service.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+//     }
+
+//     try {
+//       const data = verify( accessToken, key ) as any;
+//       console.log( accessToken, data );
+//       return next();
+//     } catch { }
+//   } );
+//   next();
+// } );
```
Comment (deshmukhmayur) [inline packages/api-gateway-service/service.ts:36]: Remove this block from this PR. When it's complete and implemented, you can send a different PR for it.

File: packages/api-gateway-service/src/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+];
+
+/* Function to fetch the public key from internal IDP*/
+// export const publicKey = () => {
+//   return fetch( process.env.INTERNAL_IDP )
+//     .then( ( res: any ) => res.json() )
+//     .then( ( res: any ) => {
+//       return res.public_key;
+//     } );
+// };
```
Comment (deshmukhmayur) [inline packages/api-gateway-service/src/helpers.ts:15]: Remove it for now. Add it back when it's implemented.

File: packages/api-gateway-service/src/types.d.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+declare module '*.graphql';declare module '*.json';
```
Comment (deshmukhmayur) [inline packages/api-gateway-service/src/types.d.ts:1]: Also, can you add make this change in the cli microservice? It seems to have come from cli template.

```suggestion
declare module '*.graphql';
declare module '*.json';
```

File: packages/api-gateway-service/webpack.common.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+const Dotenv = require('dotenv-webpack');
+const { CleanWebpackPlugin } = require('clean-webpack-plugin');
+const NodemonPlugin = require('nodemon-webpack-plugin');
+
+let envPath;
+
+if (!process.env.NODE_ENV || process.env.NODE_ENV === 'local') {
+  envPath = '.env';
+} else if (process.env.NODE_ENV === 'test') {
+  envPath = 'e2e/.test.env';
```
Comment (deshmukhmayur) [inline packages/api-gateway-service/webpack.common.js:13]: Are you sure this path works? I don't see any e2e directory in your branch.

File: packages/api-gateway-service/.env.example (hunk 1 above)
Comment (riginoommen) [reply packages/api-gateway-service/.env.example]: Included this update

File: packages/api-gateway-service/.env.example (hunk 2 above)
Comment (riginoommen) [reply packages/api-gateway-service/.env.example]: Included this update

File: packages/api-gateway-service/package.json (hunk 3 above)
Comment (riginoommen) [reply packages/api-gateway-service/package.json]: updated the tags

File: packages/api-gateway-service/service.ts (hunk 5 above)
Comment (riginoommen) [reply packages/api-gateway-service/service.ts]: removed this code

File: packages/api-gateway-service/src/helpers.ts (hunk 6 above)
Comment (riginoommen) [reply packages/api-gateway-service/src/helpers.ts]: removed this code

File: packages/api-gateway-service/src/types.d.ts (hunk 7 above)
Comment (riginoommen) [reply packages/api-gateway-service/src/types.d.ts]: formattted this code and changed this

File: packages/api-gateway-service/webpack.common.js (hunk 8 above)
Comment (riginoommen) [reply packages/api-gateway-service/webpack.common.js:13]: e2e directory and test cases are included

Comment (deshmukhmayur) [reply packages/api-gateway-service/webpack.common.js:13]: But are they working with the test env variables? Anjnee had a problem where the tests weren't using the test env values.
I think you may need to remove this test env config from here, and put it in jest.config.js

Comment (riginoommen) [reply packages/api-gateway-service/webpack.common.js:13]: Yes its working fine i tested it.

File: packages/api-gateway-service/package.json (hunk 4 above)
Comment (riginoommen) [reply packages/api-gateway-service/package.json:34]: As per todays discussion i think we can close this

**Comment by riginoommen** [comment]
Unit Tests are included
